content/6_Advanced/Vectorization.mdx

```diff
@@ -133,3 +133,275 @@ instead of faster (ex. see
 <!-- Anyways, keep in mind that these rarely make a difference in competitive programming. -->
 
 <!-- <IncompleteSection /> -->
+
+## Example - Ligatures
+
+<FocusProblem problem="tutorial" />
+
+[Problem](https://open.kattis.com/problems/ligatures?editresubmit=13798177)
+
+### Explanation
+
+The problem involves counting the occurrences of specific pairs of characters (ligatures) in a given text corpus based on multiple sets of suggested ligatures. The primary challenge is efficiently handling large input sizes and ensuring that the counting process respects the rule that ligatures can't overlap.
+
+### Solution
+
+This solution to the problem uses SIMD (Single Instruction, Multiple Data) parallelism to efficiently count the occurrences of suggested ligatures within a large corpus:
```

Comment:
i mean is there any guide you can give on how people can implement something similar themselves?

Reply:
The solution code is rather cryptic, but if you understand roughly what it is doing then you should be able to write your own solution along the same lines. Though the current solution description is rather vague so it's hard to understand what the code is doing. Here's my understanding.

The idea is to first write the naive $NQ$ solution, then figure out how to speed it up by processing batches of 256 queries in parallel. Naive solution code:

```
int main() {
	setIO();
	auto to_int = [&](char a, char b) {
		return 26 * (a - 'a') + (b - 'a');
	};  // map pair to index in [0, 676)
	def(int, N, Q, K);
	def(str, corpus);
	V<V<bool>> active(676, V<bool>(Q));
	F0R(q, Q) {
		def(str, ligature);
		for (int i = 0; i < K; ++i)
			active[to_int(ligature[2 * i], ligature[2 * i + 1])][q] = true;
	}
	F0R(q, Q) {
		int counter = 0;
		bool last = 0;
		FOR(i, 1, N) {
			last = !last & active[to_int(corpus[i - 1], corpus[i])][q];
			counter += last;
		}
		ps(counter);
	}
}
```

Now consider what needs to change if we want to process 256 queries in parallel. First, we can change `last` from a single bit to a 256-bit integer type (`__m256i` in the solution code). We can do the same with `active`.

Parallelizing the `counter += last` step is trickier. It is represented by `accs` and `ts` in the code. Basically, counter is a 20-bit integer, and we will have two 256-bit integers for each bit in counter, such that `counter = accs + ts`. To output `counter` at the end we first need to add `accs` and ts` (the `// Final accumulation across all indices` part).

Reply:
my solution code along these lines. omits optimizations such as the manual loop unrolling, but still passes just under the TL

```cpp
#pragma GCC target("avx2")
#include <immintrin.h>

using mi = __m256i;
const int SIMD_BITS = 256;

void set_bit(mi &m, size_t ind) {
	union {
		mi m;
		uint32_t ar[SIMD_BITS / 32];
	} u;
	u.m = m;
	u.ar[ind / 32] = (uint32_t)(u.ar[ind / 32] | (1 << (ind % 32)));
	m = u.m;
}

int main() {
	setIO();
	def(int, N, Q, K);
	def(str, corpus);
	vs ligs(Q);
	re(ligs);
	vi ans(Q);
	auto to_int = [&](char a, char b) { return 26 * (a - 'a') + (b - 'a'); };
	for (int ql = 0; ql < Q;) {
		int qr = min(Q, ql + SIMD_BITS);
		mi active[676]{};
		FOR(q, ql, qr)
		F0R(k, K)
		set_bit(active[to_int(ligs[q].at(2 * k), ligs[q].at(2 * k + 1))],
		        q - ql);
		mi bits_a[20]{}, bits_b[20]{};  // counter = bits_a + bits_b
		mi last{};
		FOR(i, 1, N) {
			int pair = to_int(corpus.at(i - 1), corpus.at(i));
			// update last
			last = ~last & active[pair];
			// update counter
			mi cur = last;
			for (int pos = 0;; ++pos) {  // this loop runs O(1) times on average
				if (!(i & (1 << pos))) {
					bits_b[pos] = cur;
					break;
				}
				if ((1 << (pos + 1)) > i) {
					bits_a[pos] = cur;
					break;
				}
				// reduce three bits to two (1, 1, 1 -> 1, 2)
				// bits_a, bits_b, cur -> res0, res1
				auto res0 = bits_a[pos] ^ bits_b[pos] ^ cur;
				auto res1 = (bits_a[pos] & bits_b[pos]) |
				            (cur & (bits_a[pos] | bits_b[pos]));
				bits_a[pos] = res0;
				bits_b[pos] = _mm256_setzero_si256();
				cur = res1;
			}
		}
		F0R(pos, 20) {  // extract counter from bits_a and bits_b
			union {
				mi m;
				uint32_t ar[SIMD_BITS / 32];
			} u;
			u.m = bits_a[pos];
			FOR(q, ql, qr) {
				int idx = q - ql;
				if (u.ar[idx / 32] & (1 << (idx % 32))) { ans[q] += 1 << pos; }
			}
			u.m = bits_b[pos];
			FOR(q, ql, qr) {
				int idx = q - ql;
				if (u.ar[idx / 32] & (1 << (idx % 32))) { ans[q] += 1 << pos; }
			}
		}
		ql = qr;
	}
	each(t, ans) ps(t);
	// you should actually read the stuff at the bottom
}
```

Reply:
also, some relevant resources about SIMD: https://codeforces.com/blog/entry/94152?#comment-1162758

Reply:
@Alexis-Tao88 good to link this in the internal sol